### PR TITLE
authority-discovery: futures 03 Future

### PR DESCRIPTION
**Pull request by kigawas**
#3099

### CODE REVIEW

File: core/authority-discovery/src/lib.rs [hunk 1]
```diff
@@ -42,30 +42,46 @@
...
 mod error;
 /// Dht payload schemas generated from Protobuf definitions via Prost crate in build.rs.
 mod schema {
 	include!(concat!(env!("OUT_DIR"), "/authority_discovery.rs"));
 }
 
+// TODO change tokio 02's Interval when tokio 02's runtime is adopted
+type Interval01As03 = Pin<Box<dyn Stream<Item = Instant> + Send>>;
+
+const DURATION_12H: Duration = Duration::from_secs(12 * 60 * 60);
```
Comment (tomaka) [inline core/authority-discovery/src/lib.rs]: I would strongly prefer to inline these constants.

File: core/authority-discovery/src/lib.rs [hunk 2]
```diff
@@ -291,53 +328,36 @@ where
...
 				// Make sure to call interval.poll until it returns Async::NotReady once. Otherwise, in case one of the
 				// function calls within this block do a `return`, we don't call `interval.poll` again and thereby the
 				// underlying Tokio task is never registered with Tokio's Reactor to be woken up on the next interval
 				// tick.
-				while let Async::Ready(_) = self
-					.publish_interval
-					.poll()
-					.map_err(Error::PollingTokioTimer)?
-				{}
+				while let Poll::Ready(_) = self.publish_interval.poll_next_unpin(cx) {}
```
[the remark this answers is not in the document]
Comment (kigawas) [reply core/authority-discovery/src/lib.rs]: There's no error in futures 0.3's interval

[the remark this answers is not in the document]
Comment (kigawas) [reply core/authority-discovery/src/lib.rs]: If you use futures-timer 0.4:
https://github.com/async-rs/futures-timer/blob/0.4.0/src/interval.rs#L63

If you use tokio 0.2:
https://github.com/tokio-rs/tokio/blob/master/tokio/src/timer/interval.rs#L58

**Comment by kigawas** [comment]
> I'm against using tokio_timer, as it only works if you poll using a tokio runtime. Please use `futures-timer` instead, like the rest of Substrate does. This will also remove the need for depending on `futures 0.1`.

There are still several tokio timers (mainly in grandpa) and tokio runtime is used for running tasks. Are you taking care of that?

**Comment by tomaka** [comment]
> (mainly in grandpa)

Grandpa is waiting for https://github.com/paritytech/finality-grandpa/pull/81

**Comment by mxinden** [comment]
> This will also remove the need for depending on futures 0.1.

@tomaka given that `core/service/` uses *futures 0.1* and not *futures 0.3* we don't get around depending on *futures 0.1* in the *authority discovery* module to go from *0.3* to *0.1*. Or would you prefer the conversion to happen within `core/service`?

**Comment by tomaka** [comment]
I'd prefer to do the compatibility stuff in `core/service`, so that we eliminate the dependency on futures 0.1.

**Comment by kigawas** [comment]
@tomaka 
Do you have any plan to refactor `core/network/test` to futures 0.3?

**Comment by tomaka** [comment]
@kigawas We need to update libp2p to futures 0.3 first, which is done in a separate `stable-futures` branch at the moment. We're waiting for async/await to be stable to actually finish that transition.

**Comment by kigawas** [comment]
@tomaka Oh. It'll be soon to get async/await stable (early November?)

**Comment by kigawas** [comment]
@mxinden 

I saw #3914 , can you merge this before refactoring?

**Comment by mxinden** [comment]
> I saw #3914 , can you merge this before refactoring?

Either that, or I can resolve the merge conflicts.

**Comment by mxinden** [comment]
> I'd prefer to do the compatibility stuff in core/service, so that we eliminate the dependency on futures 0.1.

If I understood @tomaka correctly, the compatibility logic should happen within `core/service`. @kigawas do you still have some spare cycles to look into this? Thanks for bearing with us!

**Comment by kigawas** [comment]
@mxinden 
No problem, I'll remove it

File: core/authority-discovery/Cargo.toml [hunk 3]
```diff
@@ -14,18 +14,17 @@ bytes = "0.4.12"
 client = { package = "substrate-client", path = "../../core/client" }
 codec = { package = "parity-scale-codec", default-features = false, version = "1.0.3" }
 derive_more = "0.15.0"
-futures = "0.1.29"
+futures03 = { package = "futures-preview", version = "0.3.0-alpha.19", features = ["compat"] }
```
Comment (tomaka) [inline core/authority-discovery/Cargo.toml]:
```suggestion
futures-preview = "0.3.0-alpha.19"
```

**Pull request merged**